---
title: 'Single-cell RNA-seq analysis using Bioconductor'
subtitle: 'Single-cell Genomics in <i class="fab fa-r-project"></i>'
author: "Kevin Rue-Albrecht"
institute: "Oxford Biomedical Data Science Training Programme"
date: "2021-03-09 (updated: `r Sys.Date()`)"
output:
  xaringan::moon_reader:
    css: [default, metropolis, rladies-fonts, "my-theme.css"]
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
# uncomment this line to produce HTML and PDF in RStudio:
knit: pagedown::chrome_print
---

layout: true

<div class="my-header"><img src="img/ox_brand1_pos.gif" alt="Oxford University Logo" align="right" height="90%"></div>

<div class="my-footer"><span>
Kevin Rue-Albrecht
&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;
Single-cell RNA-seq analysis using Bioconductor
</span></div> 

```{r setup, include = FALSE}
stopifnot(requireNamespace("htmltools"))
htmltools::tagList(rmarkdown::html_dependency_font_awesome())
knitr::opts_chunk$set(
  message = FALSE, warning = FALSE, error = FALSE,
  include = FALSE
)
options(width = 90)
stopifnot(require(tidyverse))
```

```{r, load_refs, include=FALSE, cache=FALSE}
options(htmltools.dir.version = FALSE)
library(RefManageR)
BibOptions(
  check.entries = FALSE,
  bib.style = "authoryear",
  cite.style = "authoryear",
  max.names = 2,
  style = "markdown",
  hyperlink = "to.doc",
  dashed = TRUE)
bib <- ReadBib("bibliography.bib")
```

---

# Prerequisites

<br/>

.x-large-list[
- A clone of a personal GitHub repository for this course.

- A clone of a shared GitHub repository for this course.

- A clone of <https://github.com/OBDS-Training/OBDS_environments>

- A working installation of [R](https://www.r-project.org/) (4.0.3), including the *[renv](https://rstudio.github.io/renv/articles/renv.html)* package.

- A working installation of [git](https://git-scm.com/).

- A working installation of [RStudio](https://rstudio.com/).

- Access to the CGAT cluster and the shared data folder.
]

---

# Setup

- Copy the subdirectory `bioconductor_singlecell` from the `OBDS_environments` repository to your personal <i class="fab fa-git"></i> repository for the course.

- Download the shared folder `r/bioconductor_singlecell/data` from the CGAT cluster as a subdirectory of the OBDS environment in your personal repository.

The result on your computer should look like the following

```
  obds_training/
  |_ bioconductor_singlecell/
    |_ data/
      |_ ... (data files)
```

- Launch the R project in `bioconductor_singlecell`.

- Restore the project environment using `r BiocStyle::CRANpkg("renv")`.

---

# Learning objectives

<br/>

.x-large-list[
- Work with _Bioconductor_ data structures.

- Experience a single-cell workflow using _Bioconductor_ packages.
  + Quality control (cells, features, empty droplets)
  + Normalisation
  + Clustering
  + Cluster markers (differential expression)

- Understand the relationship with other single-cell software (e.g. `r BiocStyle::CRANpkg("Seurat")`, [Scanpy](https://scanpy.readthedocs.io/en/stable/)).
]

---

# Bioconductor resources and help

```{r, include=TRUE, echo=FALSE, out.height="100px", out.width="350px", fig.align="center"}
knitr::include_graphics("https://www.bioconductor.org/images/logo/jpg/bioconductor_logo_cmyk.jpg")
```

<br/>

.x-large-list[
- Home: <https://www.bioconductor.org/>

- Support site: <https://support.bioconductor.org/>

- Courses & Conferences materials: <https://www.bioconductor.org/help/course-materials/>

- YouTube videos: <https://www.youtube.com/user/bioconductor>

- Book: [Orchestrating Single-Cell Analysis with Bioconductor](https://osca.bioconductor.org/)
]

---

# Bioconductor - Project Objectives

.center[
.x-large-text[
**Analysis and comprehension of high-throughput genomic data**
]
]

.x-large-list[
- Statistical analysis: large data, technological artifacts, designed experiments; rigorous, robust.

- Comprehension: biological context, visualization, reproducibility.

- High-throughput.

  + Sequencing: RNA-seq, ChIP-seq, variants, copy number, ...

  + Microarrays: gene expression, SNP, ...

  + Flow cytometry, proteomics, images, ...
]

???

**Source:**

- [Introduction to R / Bioconductor (2016)](https://bioconductor.org/help/course-materials/2016/BiocIntro-May/B1_Bioconductor_Intro.html) by Martin Morgan (Bioconductor Core Team)

---

# Bioconductor - Project Overview

Packages, vignettes, workflows.

- Over 1,974 software packages; but also ...

  + [Annotation packages](https://www.bioconductor.org/packages/3.10/BiocViews.html#___AnnotationData)  – static data bases of identifier maps, gene models, pathways, etc; e.g., `r BiocStyle::Biocpkg("TxDb.Hsapiens.UCSC.hg19.knownGene")`.

  + [Experiment packages](https://www.bioconductor.org/packages/3.10/BiocViews.html#___ExperimentData) – data sets used to illustrate software functionality, e.g., `r BiocStyle::Biocpkg("airway")`.

  + Discover and navigate via [biocViews](https://www.bioconductor.org/packages/devel/BiocViews.html).
  
  + Package [landing page](https://bioconductor.org/packages/release/bioc/html/Biobase.html): Title, author / maintainer, short description, citation, installation instructions, …, download statistics.

  + All user-visible functions have help pages, most with runnable examples.

  + ‘Vignettes’ an important feature in Bioconductor – narrative documents illustrating how to use the package, with integrated code, e.g. `browseVignettes("Biobase")`.

  + ‘Release’ (every six months) and ‘devel’ branches.

???

**Source:**

- [Introduction to R / Bioconductor (2016)](https://bioconductor.org/help/course-materials/2016/BiocIntro-May/B1_Bioconductor_Intro.html) by Martin Morgan (Bioconductor Core Team)

---

# Bioconductor - Philosophy

- Common data structures

  + e.g., `DNAStringSet`, `GRanges`, `GAlignemts`, `SummarizedExperiment`, `TxDb`

  + High standards of software engineering; [Core team](https://www.bioconductor.org/about/core-team/).

  + Reduce redundancy; Promotes interoperability.

  + ..., but community adoption can take time (e.g., `r BiocStyle::Biocpkg("GSEABase")`)

--

- Release cycle every 6 months (April, October)

  + Packages on the _release_ branch are stable for at least 6 months.
    Active development takes place on the _devel_ branch.
  
  + ..., but new features can take up to 6 months to be released (_devel_ $\rightarrow$ _release_).

--

- New packages are thoroughly reviewed by members of the core team

  + Packages are submitted as issues on <i class="fa fa-github"></i> [Bioconductor/Contributions](https://github.com/Bioconductor/Contributions)
  
  + Review and fixes can take time, but accepted packages are generally high quality!

???

**Source:**

- [Introduction to R / Bioconductor (2016)](https://bioconductor.org/help/course-materials/2016/BiocIntro-May/B1_Bioconductor_Intro.html) by Martin Morgan (Bioconductor Core Team)

---

# The Bioconductor SummarizedExperiment

```{r, include=TRUE, echo=FALSE, out.height="450px", fig.align='center'}
knitr::include_graphics("https://media.springernature.com/lw685/springer-static/image/art%3A10.1038%2Fnmeth.3252/MediaObjects/41592_2015_Article_BFnmeth3252_Fig2_HTML.jpg?as=webp")
```

.footnote[
**Source:** https://www.nature.com/articles/nmeth.3252 (Figure 2)
]

---

# SingleCellExperiment extends SummarizedExperiment

.pull-left[
## SummarizedExperiment.

```{r, include=TRUE, echo=FALSE, fig.align='center'}
knitr::include_graphics("https://media.springernature.com/lw685/springer-static/image/art%3A10.1038%2Fnmeth.3252/MediaObjects/41592_2015_Article_BFnmeth3252_Fig2_HTML.jpg?as=webp")
```

.small-text[
**Source:** https://www.nature.com/articles/nmeth.3252 (Figure 2)
]
]

.pull-right[
## SingleCellExperiment

```{r, include=TRUE, echo=FALSE, fig.align='center'}
knitr::include_graphics("https://raw.githubusercontent.com/Bioconductor/OSCABase/images/images/SingleCellExperiment.png")
```

.small-text[
**Source:** https://osca.bioconductor.org/data-infrastructure.html
]
]

---

# SingleCellExperiment

```{r, include=TRUE, echo=FALSE, fig.align='center'}
knitr::include_graphics("https://raw.githubusercontent.com/Bioconductor/OSCABase/images/images/SingleCellExperiment.png")
```

.footnote[
**Source:** https://osca.bioconductor.org/data-infrastructure.html
]

---

# A single-cell analysis workflow using Bioconductor

.pull-left[
```{r, include=TRUE, echo=FALSE, fig.align='center', out.height='500px'}
knitr::include_graphics("https://raw.githubusercontent.com/Bioconductor/OSCABase/images/images/Workflow.png")
```
]

.pull-right[
**Source:** https://osca.bioconductor.org/overview.html#data-processing-and-downstream-analysis

<br/>

**Key points:**

- A shared data structure allows modular analysis.
- Individual packages provide
  + Data structures (`r BiocStyle::Biocpkg("SingleCellExperiment")`)
  + Method (`r BiocStyle::Biocpkg("scran")`, `r BiocStyle::Biocpkg("scater")`)
  + Annotation (`r BiocStyle::Biocpkg("org.Hs.eg.db")`)
  + Reporting (`r BiocStyle::CRANpkg("rmarkdown")`, `r BiocStyle::CRANpkg("shiny")`)
]

---

# Bioconductor and Seurat

.pull-left[
## Bioconductor (2001-)

<br/>

![Schematic representation of the Bioconductor SingleCellExperiment](https://raw.githubusercontent.com/Bioconductor/OSCABase/images/images/SingleCellExperiment.png)

.small-text[
**Source:** https://osca.bioconductor.org/data-infrastructure.html
]
]

.pull-right[
## Seurat (2014-)

<br/>

.xx-small-table[
|Slot         |Function                                                                        |
|:------------|:-------------------------------------------------------------------------------|
|assays       |A list of assays within this object                                             |
|meta.data    |Cell-level meta data                                                            |
|active.assay |Name of active, or default, assay                                               |
|active.ident |Identity classes for the current object                                         |
|graphs       |A list of nearest neighbor graphs                                               |
|reductions   |A list of DimReduc objects                                                      |
|project.name |User-defined project name (optional)                                            |
|tools        |Empty list. Tool developers can store any internal data from their methods here |
|misc         |Empty slot. User can store additional information here                          |
|version      |Seurat version used when creating the object                                    |
]

.small-text[
**Source:** https://github.com/satijalab/seurat/wiki/Seurat
]
]

---

# Bioconductor and Scanpy

.pull-left[
## Bioconductor (2001-)

<br/>

```{r, include=TRUE, echo=FALSE, fig.align='center'}
knitr::include_graphics("https://media.springernature.com/lw685/springer-static/image/art%3A10.1038%2Fnmeth.3252/MediaObjects/41592_2015_Article_BFnmeth3252_Fig2_HTML.jpg?as=webp")
```

.small-text[
**Source:** https://www.nature.com/articles/nmeth.3252 (Figure 2)
]
]

.pull-right[
## Scanpy (2017-)

<br/>

![Schematic representation of the Scanpy AnnData](img/anndata.svg)

.small-text[
**Source:** [https://anndata.readthedocs.io](https://anndata.readthedocs.io/en/latest/anndata.AnnData.html)
]
]

---

# Import single-cell data using Bioconductor

The package `r BiocStyle::Biocpkg("DropletUtils")` provides several utilities for reading and preprocessing single-cell data.

## Import

- `read10xCounts()` imports 10X Genomics data from the CellRanger output directories.
  It returns a `SingleCellExperiment` object.

## Utilities and quality control

- `barcodeRanks()` ccomputes barcode rank statistics and identify the knee and inflection points on the total count curve.
- `emptyDrops()` is designed to distinguish between empty droplets and cells by testing each barcode’s expression profile for significant deviation from the ambient profile.
- `swappedDrops()` removes the effects of barcode swapping.
  It takes a list of paths to individual samples and returns a list with an item called `cleaned`, itself a list of sparse matrices, after removing swapped molecules.

---

# Empty droplets

.pull-left[
```{r, include=TRUE, echo=FALSE, out.height='500px'}
knitr::include_graphics("https://cores.research.asu.edu/sites/default/files/inline-images/10x-barcoded%20gel%20beads.png")
```
]

.pull-right[
.x-large-list[
- To optimise the capture rate, the number of GEL beads is much higher than the number of cells.

- This leads to a higher probability of a cell and a bead ending up in the same droplet.

- However, this also leads to a large number of 'empty' droplets with just a bead and some ambient RNA or a dead cell.
]
]

---

# Empty droplets

Below is an example of a challenging cell-calling scenario mixing:

- 300 high RNA content 293T cells
- 2000 low RNA content PBMC cells

On the left is the cell calling result with the cell calling algorithm prior to Cell Ranger 3.0 and on the right is the current Cell Ranger 3.0 result. You can see that low RNA content cells are successfully identified by the new algorithm.

.pull-left[
```{r, include=TRUE, echo=FALSE}
knitr::include_graphics("https://support.10xgenomics.com/img/single-cell-gex/knee-plot-old-cell-calling.png")
```
]

.pull-right[
```{r, include=TRUE, echo=FALSE}
knitr::include_graphics("https://support.10xgenomics.com/img/single-cell-gex/knee-plot-new-cell-calling.png")
```
]

---

# Identifying empty droplets

The `r BiocStyle::Biocpkg("DropletUtils")` package includes the `emptyDrops()` function.

.pull-left[
## Parameters

- `m`, a matrix of UMI counts

- `lower`, total UMI count below which barcodes are assumed to be empty droplets.

- `niters`, number of iterations to use for the Monte Carlo p-value calculations.
  More interations increase the sensitivity to yield lower $p$-values and identify more cell-containing droplets, at the cost of computing time.
]


.pull-right[
## Method
- Barcodes with fewer than `lower` are used to estimate the composition of ambient RNA.

- The count vector for each barcode above lower is then tested for a significant deviation from these proportions.

- $p$-values are corrected for multiple testing using the Benjamini-Hochberg correction.
]

## Output

A table including the probability that each cell is an empty droplet.

---

# Doublets

.pull-left[
```{r, include=TRUE, echo=FALSE, out.height='500px'}
knitr::include_graphics("https://cores.research.asu.edu/sites/default/files/inline-images/10x-barcoded%20gel%20beads.png")
```
]

.pull-right[
.x-large-list[
- Occasionally, two or more cells are captured in the same droplet.

Estimated doublets rates are:

| Doublet rate (%)| Estimated cells|
|----------------:|---------------:|
|              0.8|            1000|
|              1.6|            2000|
|              2.3|            3000|
|              3.1|            4000|
|              3.9|            5000|
]
]

---

# Identifying doublets using artificial doublets

.pull-left[
`r BiocStyle::Githubpkg("chris-mcginnis-ucsf/DoubletFinder")` (GitHub) and `r BiocStyle::Biocpkg("scDblFinder")` (Bioconductor) take a similar approach.

## Steps

1.  Generate artificial doublets from existing scRNA-seq data

2. Pre-process merged real-artificial data

3. Perform PCA and use the PC distance matrix to find each cell's proportion of artificial k nearest neighbors (pANN)

4. Rank order and threshold pANN values according to the expected number of doublets

<i class="fab fa-github"></i> [chris-mcginnis-ucsf/DoubletFinder](https://github.com/chris-mcginnis-ucsf/DoubletFinder)
]

.pull-right[
```{r, include=TRUE, echo=FALSE, fig.align='center', out.height='450px'}
knitr::include_graphics("https://marlin-prod.literatumonline.com/cms/attachment/11ab5a3c-8f8a-4e8d-a333-7e9feedb846a/fx1_lrg.jpg")
```

`r Citet(bib, "mcginnis2019doubletfinder")`
]

---

# Summary of doublet detection methods

## GitHub

- `r BiocStyle::Githubpkg("chris-mcginnis-ucsf/DoubletFinder")`, the original method using artificial doublets.

## Bioconductor

- `r BiocStyle::Biocpkg("scDblFinder")`, a Bioconductor implementation of `DoubletFinder`. Faster, single command.

- `scran::doubletCells()` identifies potential doublet cells based on simulations of putative doublet expression profiles.

---

# Quality-control and preprocessing using Bioconductor

The package `r BiocStyle::Biocpkg("scater")` provides several utilities focused on the preprocessing and quality control of single-cell gene expression data.

## Quality-control

- `perCellQCMetrics()` and `perFeatureQCMetrics()` compute a variety of quality control metrics for arbitrary sets of cells and features in the dataset (e.g., mitochondrial genes, individual samples).

## Preprocessing

- `logNormCounts()` computes log-transformed normalized expression values.
  This includes the computation of size factors using `librarySizeFactors()`.

---

# scran: Methods for single-cell RNA-seq data analysis

The package `r BiocStyle::Biocpkg("scran")` provides a range of methods focused on the normalisation and preprocessing of single-cell gene expression data.
Some examples are presented below.

.pull-left[
### Normalisation

- `computeSumFactors()` implements the deconvolution strategy from (Lun et al., 2016).

### Feature selection

- `modelGeneVar()` decomposes the log-expression profiles for each gene into technical and biological components based on a fitted mean-variance trend.
- `getTopHVGs()` uses modelling statistics `modelGeneVar()` to define a set of highly variable genes.
]

.pull-right[
### Clustering

- `quickCluster()` provides a convenient wrapper to quickly define clusters of a minimum size.

### Differential expression

- `findMarkers()` finds candidate marker genes for groups of cells (e.g., clusters).
]

---

# scran: Deconvoluting size factors from cell pools

.pull-left[
- All cells in the data set are averaged to make a reference pseudo-cell.

- Expression values for cells in pool A are summed together and normalized against the reference to yield a pool-based size factor.

- This is equal to the sum of the cell-based factors and can be used to formulate a linear equation.

- Repeating this for multiple pools leads to the construction of a linear system that can be solved to estimate the size factor for each cell
]

.pull-right[
```{r, include=TRUE, echo=FALSE}
knitr::include_graphics("https://media.springernature.com/full/springer-static/image/art%3A10.1186%2Fs13059-016-0947-7/MediaObjects/13059_2016_947_Fig3_HTML.gif?as=webp")
```

.right[
`r Citet(bib, "lun2016pooling")`
]
]

---

# Dimensionality reduction in Bioconductor

The package `r BiocStyle::Biocpkg("scater")` provides wrappers for several common dimensionality reduction methods.
Those include:

.x-large-list[
- `runPCA()`
- `runTSNE()`
- `runUMAP()`
]

## Important notes:

- Those methods generally use only a subset most variable features, not the entire dataset.
- By default, `runTSNE()` performs a PCA step first, the result of which the first 50 PCs are used as input for the t-SNE algorithm.
- By default, `runUMAP()` performs a PCA step first, of which the first 50 PCs are used as input into the UMAP algorithm.

---

# Cluster cells

The package `r BiocStyle::Biocpkg("scran")` provides several methods for clustering single-cell data.

- `quickCluster()` clusters similar cells using either log-expression values or ranks.
- `buildSNNGraph()` builds a shared nearest-neighbour graph using cells as nodes.
  For each cell, its $k$ nearest neighbours are identified using the `findKNN()` function, based on distances between their expression profiles (Euclidean by default).
  + The function works on a reduced dimension representation of the dataset, _not_ the gene expression matrix.
  + The output is a graph where nodes are cells and edges represent connections between nearest neighbors.
    To obtain cluster, the graph must still be partitioned using an appropriate function, e.g. `igraph::cluster_louvain()`.

## Example

```{r, include=TRUE, eval=FALSE}
g <- buildSNNGraph(sce, use.dimred = 'PCA')
colData(sce)[["cluster_louvain"]] <- factor(igraph::cluster_louvain(g)$membership)
```

---

# Cluster markers

The package `r BiocStyle::Biocpkg("scran")` provides several methods for identifying cluster markers.

- `findMarkers()` tests for differential expression between pairs of groups.

- `pairwiseTTests()` performs pairwise Welch t-tests between groups of cells, possibly after blocking on uninteresting factors of variation.
  The return value is a list of `DataFrame` - one for each group of cells - that can be examined separately of combined into a single table using `combineMarkers()`.

- `pairwiseWilcox()` and `pairwiseBinom()` provide alternative statistical tests.

---

# More single-cell methods available in Bioconductor

## scater

- `r BiocStyle::Biocpkg("scater")` includes a variety of convenient plotting functions (e.g. `plotReducedDim()`, `plotHeatmap()`)

## scran

- `scran::bootstrapCluster()` generates bootstrap replicates and reclusters on them to determine the stability of clusters with respect to sampling noise.
- `scran::cyclone()` implements the cell cycle classification step of the prediction method described by Scialdone et al. (2015).

---

# iSEE - interactive SummarizedExperiment Explorer

```{r, include=TRUE, echo=FALSE, fig.align='center', out.height='500px'}
knitr::include_graphics("https://f1000researchdata.s3.amazonaws.com/manuscripts/16293/6bf85f9d-8352-4a78-a8da-456f05f5c4c9_figure1.gif")
```

---

# iSEE - interactive SummarizedExperiment Explorer

```{r, include=TRUE, eval=FALSE}
library(iSEE)
library(scRNAseq)

# Example data ----
sce <- ReprocessedAllenData(assays="tophat_counts")
class(sce)

library(scater)
sce <- logNormCounts(sce, exprs_values="tophat_counts")

sce <- runPCA(sce, ncomponents=4)
sce <- runTSNE(sce)
rowData(sce)$ave_count <- rowMeans(assay(sce, "tophat_counts"))
rowData(sce)$n_cells <- rowSums(assay(sce, "tophat_counts") > 0)
sce

# launch the app itself ----

app <- iSEE(sce)
shiny::runApp(app, port=1234)
```

---

# iSEE - interactive SummarizedExperiment Explorer

.center[`iSEE(sce, voice=TRUE)`]

<div style="position: relative; padding-bottom: 56.25%; height: 0; overflow: hidden;">
  <iframe src="https://www.youtube.com/embed/0crFZLwAJOE?autoplay=0" style="position: absolute; top: 0; left: 0; width: 100%; height: 90%; border:0;" allowfullscreen title="YouTube Video"></iframe>
</div>

.footnote[
<https://youtu.be/0crFZLwAJOE>
]

---

# Exercise - A typical workflow using Bioconductor

1. Import the filtered matrix into R; use `r BiocStyle::Biocpkg("DropletUtils")`.
  What can you tell about the raw data?

2. Compute and visualise quality control metrics; use `r BiocStyle::Biocpkg("scater")`.
  Would you remove any cell?

3. Convert the counts into normalized expression values to eliminate cell-specific biases (e.g., in capture efficiency); use `r BiocStyle::Biocpkg("scater")` and/or `r BiocStyle::Biocpkg("scran")`.

4. Select features for downstream analyses, e.g. highly variable genes; use `r BiocStyle::Biocpkg("scran")`.

5. Apply dimensionality reduction to compact the data and further reduce noise; use `r BiocStyle::Biocpkg("scater")`.

6. Cluster cells; use `r BiocStyle::Biocpkg("scran")`.

7. Identify markers for each cluster; use `r BiocStyle::Biocpkg("scran")`.

8. At every step, take the time to _visualise_ results and relate them to the raw/normalised data; try `r BiocStyle::Biocpkg("iSEE")`.

9. Bonus point: Import the raw matrix, and:

  + Identify and exclude empty droplets; use `r BiocStyle::Biocpkg("DropletUtils")`.

  + Identify doublets; use `r BiocStyle::Biocpkg("scDblFinder")`.

???

Aim for <https://osca.bioconductor.org/overview.html#quick-start>

---

# Exercise step 1. Import the filtered matrix

- Import the filtered matrix into R; use `r BiocStyle::Biocpkg("DropletUtils")`.

**Note:** use the `samples` argument of the `DropletUtils::read10xCounts()` function to give a convenient name to each sample.
  Check the difference without using the `samples` argument.

- Print a summary representation of the object.
  What can you tell about the contents of the object?

- What can you tell from the object metadata?

**Note:** slots of `SummarizedExperiment` objects are typically accessed using functions of the same name, e.g. `metadata()`.

---

# Exercise step 2. Quality control

- Compute and visualise quality control metrics; use `r BiocStyle::Biocpkg("scater")`.

**Note:** identify mitochondrial genes and pass those to the `subsets` argument of the `scater::addPerCellQC()` function.

**Note:** what is the return value?
  where are the quality metrics stored?
  what is the difference with `scater::perCellQCMetrics()`?

- Visualise quality metrics; use `r BiocStyle::CRANpkg("ggplot2")`.

- Would you remove any cell?

**Note:** we have imported a matrix prefiltered by [Cell Ranger](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger) using a certain set of criteria (which ones?); what other criteria can you think of, which may be useful for filterering cells?

- Similarly, use `scater::perFeatureQCMetrics()` or `scater::addPerFeatureQC()` to compute per-feature quality metrics, and visualise those metrics.

---

# Exercise step 3. Normalisation

- Convert the counts into normalized expression values to eliminate cell-specific biases (e.g., in capture efficiency); use `r BiocStyle::Biocpkg("scater")` and/or `r BiocStyle::Biocpkg("scran")`.

**Note:** use `scater::logNormCounts()` to compute log-normalised counts.
  What is the return value?
  Where can you find the normalised counts?

- Plot the variance against the mean of each gene.

**Note:** how can you tell whether the normalisation was effective?
  Compare with https://osca.bioconductor.org/feature-selection.html#quantifying-per-gene-variation

---

# Exercise step 4. Feature selection

Select features for downstream analyses, e.g. highly variable genes; use `r BiocStyle::Biocpkg("scran")`.

- Use `scran::modelGeneVar()` to model the variance of the log-expression profiles for each gene.
  What is the output?

- Visualise the relation between the mean expression of each gene and the total / biological / technical variance of each gene.

**Note:** How do you interpret those different values?

- Use `scran::getTopHVGs()` to identify highly variable genes.

**Note:** what is the output?
  How many genes do you identify?
  Where are those genes located in the mean-variance plot?
  What happens to this plot if you set more stringent thresholds to define highly variable genes?

---

# Exercise step 5. Dimensionality reduction

- Use `scater::runPCA` to compact the data and further reduce noise.

**Note:** only give the set of highly variable genes to the `scater::runPCA()` function, to save time, memory, and to focus on biologically variable genes in the data set.

- Continue with UMAP or t-SNE.

**Note:** UMAP and t-SNE are typically given the output of PCA, to further reduce the dimensionality for ease of visualisation.

- Visualise the layout of cells produced by each of those dimensionality reduction methods.
  Considering coloring points with quality control metrics.

## Bonus point

- use `scran::denoisePCA()` to remove principal components that correspond to technical noise, and compare downstream t-SNE or UMAP with those obtained before de-noising.

---

# Exercise step 6. Clustering

- Use `scran::getClusteredPCs()` to cluster cells after using varying number of PCs and pick the number of PCs using a heuristic based on the number of clusters.

- Use `scran::buildSNNGraph()` and `igraph::cluster_louvain()` with the "ideal" number of PCs.

- Visualise the assigned cluster on your preferred dimensionality reduction layout.

**Note:** Dimensionality reduction and clustering are two separate methods both based on the PCA output.
  They may not always agree with each other, often helping to diagnose over- or under-clustering.

## Bonus point

- Test different numbers of principal components and compare results.

- Try `scran::quickCluster()`; identify key parameters and compare results.

---

# Exercise step 7. Cluster markers

- Use `scran::findMarkers()` to identify markers for each cluster.

- Visualise the expression of selected markers:

  + as a dot plot, optionally with a violin layer.
  
  + on a dimensionality reduction layout.
  
---

# Exercise step 8. iSEE

- Use `iSEE::iSEE()` to launch an interactive web-application to visualise the contents of the `SingleCellExperiment` object.

## Bonus point

- Preconfigure the application to start with a subset of panels, e.g.

```{r, eval=FALSE, include=TRUE}
iSEE::iSEE(sce, initial = list(
  ReducedDimensionPlot(PanelWidth=4L),
  RowDataTable(PanelWidth=8L)
))
```

---

# Further reading

- [Introduction to R / Bioconductor (2016)](https://bioconductor.org/help/course-materials/2016/BiocIntro-May/B1_Bioconductor_Intro.html) by Martin Morgan (Bioconductor Core Team)

- [HarvardX Biomedical Data Science Open Online Training](http://rafalab.github.io/pages/harvardx.html)

---

# References

.small-text[
```{r, include=TRUE, echo=FALSE, results="asis"}
PrintBibliography(bib)
```
]
