# The origin of empty droplets

.pull-left[
```{r, include=TRUE, echo=FALSE, out.height='500px', out.width='300px'}
# Source: https://cores.research.asu.edu/sites/default/files/inline-images/10x-barcoded%20gel%20beads.png
knitr::include_graphics("img/10x-barcoded-gel-beads.png")
```
]

.pull-right[
- To optimise the capture rate, the number of GEL beads is much higher than the number of cells.

- This leads to a higher probability of a cell and a bead ending up in the same droplet.

- However, this also leads to a large number of 'empty' droplets with just a bead and some ambient RNA or a dead cell.

- But also, droplets can occasionally capture more than one cell.

  + 1,000 cells ~ 0.8% doublets
  + 2,000 cells ~ 1.6% doublets
  + 3,000 cells ~ 2.3% doublets
  + 4,000 cells ~ 3.1% doublets
  + 5,000 cells ~ 3.9% doublets
  + etc ...
]

---

# Empty droplets contain fewer transcripts

Below is an example of a challenging cell-calling scenario mixing:

- 300 high RNA content 293T cells
- 2000 low RNA content PBMC cells

On the left is the cell calling result with the cell calling algorithm prior to Cell Ranger 3.0 and on the right is the current Cell Ranger 3.0 result. You can see that low RNA content cells are successfully identified by the new algorithm.

.pull-left[
```{r, include=TRUE, echo=FALSE}
# Source: https://support.10xgenomics.com/img/single-cell-gex/knee-plot-old-cell-calling.png
knitr::include_graphics("img/knee-plot-old-cell-calling.png")
```
]

.pull-right[
```{r, include=TRUE, echo=FALSE}
# Source: https://support.10xgenomics.com/img/single-cell-gex/knee-plot-new-cell-calling.png
knitr::include_graphics("img/knee-plot-new-cell-calling.png")
```
]

.small-p[
Source: <https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/algorithms/overview>
]

---

# The emptyDrops method

The function `DropletUtils::emptyDrops()` estimates the probability of empty droplets.

.pull-left[
## Parameters

- `m`, a matrix of UMI counts

- `lower`, total UMI count below which barcodes are assumed to be empty droplets.

- `niters`, number of iterations to use for the Monte Carlo p-value calculations.
  More interations increase the sensitivity to yield lower $p$-values and identify more cell-containing droplets, at the cost of computing time.
]


.pull-right[
## Method
- Barcodes with fewer than `lower` are used to estimate the composition of ambient RNA.

- The count vector for each barcode above lower is then tested for a significant deviation from these proportions.

- $p$-values are corrected for multiple testing using the Benjamini-Hochberg correction.
]

## Output

A table including the probability that each cell is an empty droplet.

---

# Exercise

## Empty droplets

- Import the raw matrix into <i class="fab fa-r-project"></i>; use `r BiocStyle::Biocpkg("DropletUtils")`.

- Remove droplets without any count at all.

This will massively speed up downstream steps, while removing unquestionable empty droplets.

- Run `DropletUtils::emptyDrops()`.

**Note:**
  How do you read the output?
  Which droplets would you consider empty?

- Draw plots comparing empty droplets and other droplets
  (e.g., library size, mitochondrial content, PCA).

- Import the filtered matrix for the same cells and compare their set of filtered cells to yours.

- Remove empty droplets from the data sets.

---

# The origin of doublets

.pull-left[
```{r, include=TRUE, echo=FALSE, out.height='500px', out.width='300px'}
# Source: https://cores.research.asu.edu/sites/default/files/inline-images/10x-barcoded%20gel%20beads.png
knitr::include_graphics("img/10x-barcoded-gel-beads.png")
```
]

.pull-right[
- To optimise the capture rate, the number of GEL beads is much higher than the number of cells.

- This leads to a higher probability of a cell and a bead ending up in the same droplet.

- However, this also leads to a large number of 'empty' droplets with just a bead and some ambient RNA or a dead cell.

- But also, droplets can occasionally capture more than one cell.

  + 1,000 cells ~ 0.8% doublets
  + 2,000 cells ~ 1.6% doublets
  + 3,000 cells ~ 2.3% doublets
  + 4,000 cells ~ 3.1% doublets
  + 5,000 cells ~ 3.9% doublets
  + etc ...
]

---

# The DoubletFinder method

.pull-left[
## Steps

1. Generate artificial doublets from existing scRNA-seq data.

2. Pre-process merged real-artificial data.

3. Perform PCA and use the PC distance matrix to find each cell's proportion of artificial k nearest neighbors (pANN).

4. Rank order and threshold pANN values according to the expected number of doublets.
]

.pull-right[
```{r, include=TRUE, echo=FALSE, fig.align='center', out.height='450px'}
## Source: https://www.sciencedirect.com/science/article/pii/S2405471219300730
knitr::include_graphics("img/doubletfinder-fig1.jpeg")
```
]

.footnote[
`r Citet(bib, "mcginnis2019doubletfinder")`

<i class="fab fa-github"></i> [chris-mcginnis-ucsf/DoubletFinder](https://github.com/chris-mcginnis-ucsf/DoubletFinder)
]

---

# DoubletFinder in a workflow

.pull-left[
- Works on `r BiocStyle::CRANpkg("Seurat")` objects.

- Preprocess the Seurat object

  + Normalise, scale, variable features, dimensionality reduction.

- Sweep combinations of parameters pN and pK. identify optimal pK parameter

- Estimate proportion of homotypic doublets

- Classify / predict / call doublets.
]

.pull-right[
```{r, include=TRUE, echo=FALSE, fig.align='center'}
## Source: https://www.sciencedirect.com/science/article/pii/S2405471219300730
knitr::include_graphics("img/doubletfinder-fig2.jpeg")
```
]

.footnote[
`r Citet(bib, "mcginnis2019doubletfinder")`

<i class="fab fa-github"></i> [chris-mcginnis-ucsf/DoubletFinder](https://github.com/chris-mcginnis-ucsf/DoubletFinder)
]

---

# DoubletFinder parameters

| Name | Description |
|:-----|:------------|
| `seu` | A fully-processed Seurat object (i.e., After NormalizeData, FindVariableGenes, ScaleData, and RunPCA have all been performed). |
| `PCs` | The number of (significant) PCs to use for computing the distance matrix. |
| `pN` | The proportion of artificial doublets to generate, relative to the number of cells in the dataset. |
| `pK` | The proportion of cells in the merged dataset to use as nearest neighbors when testing each cell. |
| `nExp` | The total number of doublets expected in the dataset.<br/>This value can best be estimated from cell loading densities into the 10X/Drop-Seq device, and adjusted according to the estimated proportion of homotypic doublets. |

See `?DoubletFinder::doubletFinder()` for more details.

---

# The scDblFinder method

The rough logic is very similar to other methods (e.g. `r BiocStyle::Githubpkg("chris-mcginnis-ucsf/DoubletFinder")`, `r BiocStyle::Biocpkg("scds")`), with a few twists that make it more efficient and provide extra features `r Citep(bib, "germain2020scdblfinder")`.

## Parameters

| Name | Description |
|:-----|:------------|
| `sce` | A `SummarizedExperiment` or count matrix. |
| `clusters` | Optional cluster assignments (if omitted, will run clustering). |
| `samples` | Vector indicating which sample each cell belong to. Samples are understood to be processed independently. |
| `artificialDoublets` | Approximate number of doublets to identify (if omitted, will be $5 \times nbClusters^2$). |
| `k` | Number of nearest neighbours for KNN graph. |

.footnote[
Bioconductor: `r BiocStyle::Biocpkg("scDblFinder")`.
See `?scDblFinder::scDblFinder()` for more details.
]

---

# Other Doublet Detection Methods

## R

.x-large-list[
- `r BiocStyle::Biocpkg("scds")` (Bioconductor)

- <i class="fab fa-github"></i> [EDePasquale/DoubletDecon](https://github.com/EDePasquale/DoubletDecon)
  `r Citep(bib, "depasquale2019doubletdecon")`
]

## Python

.x-large-list[
- <i class="fab fa-github"></i> [JonathanShor/DoubletDetection](https://github.com/JonathanShor/DoubletDetection)

- <i class="fab fa-github"></i> [Scrublet](https://github.com/AllonKleinLab/scrublet)
  `r Citep(bib, "wolock2019scrublet")`
]

---

# Exercise

## Doublets

Continue using the data set after you excluded empty droplets.

- Run `scDblFinder::scDblFinder()`.

- Draw plots comparing doublets and other droplets.
  (e.g., library size, mitochondrial content, PCA)

- Remove doublets from the data set.

## Bonus point

- Compare results with the output of `r BiocStyle::Githubpkg('chris-mcginnis-ucsf/DoubletFinder')`

  + You will need to make a `Seurat` object.

---

# The origin of ambient RNA

.pull-left[
```{r, include=TRUE, echo=FALSE, fig.align='center'}
# Source: https://twitter.com/GenomeBiology/status/1235571002903576576/photo/1
knitr::include_graphics("img/ambient_rna_origin.png")
```

`r Citet(bib, "yang2020decontx")`
]

.pull-right[
- Cells damaged during library preparation can release their cytoplasmic contents (e.g., RNA transcripts) into the ambient medium.

- In droplet-based microfluidic devices, ambient RNA can be incorporated into droplets along with oligonucleotide-barcoded beads and cells.

- Both native mRNA from the cell and contaminating ambient RNA will be barcoded and counted within a droplet.
]

<br/>

## Objective

Deconvolute ambient RNA signal from cellular transcripts to estimate true biological signal as precisely as possible.

---

# The decontX method

.pull-left[
```{r, include=TRUE, echo=FALSE, fig.align='center', out.height='450px'}
# Source: https://media.springernature.com/lw685/springer-static/image/art%3A10.1186%2Fs13059-020-1950-6/MediaObjects/13059_2020_1950_Fig1_HTML.png
knitr::include_graphics("img/decontx_Fig1_HTML.png")
```

`r Citet(bib, "yang2020decontx")`
]

.pull-right[
- Bioconductor package `r BiocStyle::Biocpkg("celda")`

- Function `celda::decontX()`

- **Only
the expression profile of "real" cells after cell calling (i.e., removing empty droplets and doublets) are required to run DecontX.**

- **Empty cell droplet information (low expression cell barcodes before cell calling) are
not needed.**
]

---

# The decontX method

The function `celda::decontX()` identifies contamination from factors such as ambient RNA in single cell genomic datasets.
It estimate matrices of native expression and contamination from ambient RNA.

.pull-left[
## Parameters

- `x`, a `SingleCellExperiment` object.

- `z`, broad cell cluster labels.

- `batch`, batch labels.

- `maxIter`, maximum iterations for the EM algorithm.
]


.pull-right[
## Method

Assumption: each cell is a mixture of two multinomial distributions:

1. A distribution of native transcripts from the cellâ€™s true population

2. A distribution of contaminating transcripts from all other cell populations captured in the assay
]

## Output

A `SingleCellExperiment` object with the decontaminated counts stored as a new assay.
Other pieces of information are stored in `colData()` and `metadata()`.
See `?celda::decontX()` for more information.

---

# Exercise

## Ambient RNA

- Run `celda::decontX()` (continue using the data after excluding empty droplets and doublets).

- Visualize the UMAP layout created by `decontX`; use `plotDimReduceCluster()`.

- Visualize the percentage of contamination estimated by `decontX` in each cell.
  Use the `plotDecontXContamination()` function.

.center[
**Does any cluster display higher (estimated) levels of ambient RNA?**
]

- Visualize the log-normalised expression of marker genes on a UMAP layout faceted by gene; use `plotDimReduceFeature()`.

**Note:** For instance, use _CD3D_, _CD3E_, _GNLY_, _LYZ_, _S100A8_, _S100A9_, _CD79A_, _CD79B_, and _MS4A1_.

- Visualise the percentage of cells with detectable expression of marker genes before and after decontamination as a bar plot; use `plotDecontXMarkerPercentage()`.

- Visualize the log-normalised expression of marker genes before and after decontamination as a violin plot; use `plotDecontXMarkerExpression()`.

---

# References

.small-p[
```{r refs, include=TRUE, echo=FALSE, results="asis"}
PrintBibliography(bib)
```
]
